AMDGPU/GlobalISel: AMDGPURegBankLegalize

petar-avramovic opened the pull request:
Lower G_ instructions that can't be inst-selected with register bank
assignment from AMDGPURegBankSelect based on uniformity analysis.
- Lower instruction to perform it on assigned register bank
- Put uniform value in vgpr because SALU instruction is not available
- Execute divergent instruction in SALU - "waterfall loop"

Given LLTs on all operands after legalizer, some register bank
assignments require lowering while other do not.
Note: cases where all register bank assignments would require lowering
are lowered in legalizer.

AMDGPURegBankLegalize goals:
- Define Rules: when and how to perform lowering
- Goal of defining Rules it to provide high level table-like brief
  overview of how to lower generic instructions based on available
  target features and uniformity info (uniform vs divergent).
- Fast search of Rules, depends on how complicated Rule.Predicate is
- For some opcodes there would be too many Rules that are essentially
  all the same just for different combinations of types and banks.
  Write custom function that handles all cases.
- Rules are made from enum IDs that correspond to each operand.
  Names of IDs are meant to give brief description what lowering does
  for each operand or the whole instruction.
- AMDGPURegBankLegalizeHelper implements lowering algorithms

Since this is the first patch that actually enables -new-reg-bank-select
here is the summary of regression tests that were added earlier:
- if instruction is uniform always select SALU instruction if available
- eliminate back to back vgpr to sgpr to vgpr copies of uniform values
- fast rules: small differences for standard and vector instruction
- enabling Rule based on target feature - salu_float
- how to specify lowering algorithm - vgpr S64 AND to S32
- on G_TRUNC in reg, it is up to user to deal with truncated bits
  G_TRUNC in reg is treated as no-op.
- dealing with truncated high bits - ABS S16 to S32
- sgpr S1 phi lowering
- new opcodes for vcc-to-scc and scc-to-vcc copies
- lowering for vgprS1-to-vcc copy (formally this is vgpr-to-vcc G_TRUNC)
- S1 zext and sext lowering to select
- uniform and divergent S1 AND(OR and XOR) lowering - inst-selected into
  SALU instruction
- divergent phi with uniform inputs
- divergent instruction with temporal divergent use, source instruction
  is defined as uniform(AMDGPURegBankSelect) - missing temporal
  divergence lowering
- uniform phi, because of undef incoming, is assigned to vgpr. Will be
  fixed in AMDGPURegBankSelect via another fix in machine uniformity
  analysis.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalize.cpp:
```diff
@@ -69,6 +81,241 @@ FunctionPass *llvm::createAMDGPURBLegalizePass() {
+    }
+  }
+
+  assert(!hasSGPRS1(MF, MRI) && "detetected SGPR S1 register");
```
Typo detetected

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalize.cpp:
```diff
@@ -69,6 +81,241 @@ FunctionPass *llvm::createAMDGPURBLegalizePass() {
+          continue;
+        }
+
+        llvm_unreachable("missing anyext + trunc combine\n");
```
Don't need newline in unreachable error strings

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeHelper.h:
```diff
@@ -0,0 +1,118 @@
+  const RegisterBank *VgprRB;
+  const RegisterBank *VccRB;
+
+  LLT S1 = LLT::scalar(1);
```
static constexpr

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeRules.cpp:
```diff
@@ -0,0 +1,334 @@
+  case DivP1:
+    return MRI.getType(Reg) == LLT::pointer(1, 64) && MUI.isDivergent(Reg);
+
+  case _:
```
Huh?

petar-avramovic replied:
That is "don't check anything", for example used in:
```
  addRulesForIOpcs({amdgcn_frexp_exp})
      .Any({{DivS32, _, S32}, {{Vgpr32}, {IntrId, Vgpr32}}})
      .Any({{DivS32, _, S64}, {{Vgpr32}, {IntrId, Vgpr64}}});
```
it is used to skip non-virtual-registers(intrinsics ID, imm, physical register or $noreg) in PredicateMapping when writing rules for amdgpu-intrinsics

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeHelper.h:
```diff
@@ -0,0 +1,118 @@
+  Register createVcc() { return MRI.createVirtualRegister({VccRB, S1}); }
+
+  const RegisterBank *getRegBank(Register Reg) {
+    const RegisterBank *RB = MRI.getRegBankOrNull(Reg);
+    // This assert is not guaranteed by default. RB-select ensures that all
+    // instructions that we want to RB-legalize have reg banks on all registers.
+    // There might be a few exceptions. Workaround for them is to not write
+    // 'mapping' for register operand that is expected to have reg class.
+    assert(RB);
+    return RB;
```
Should just introduce an asserting variant directly in MRI. What about the assigned register class case?

petar-avramovic replied:
> What about the assigned register class case?

I intended for RBLegalize to work with register banks only.
RBSelect is meant to deal with assigned register classes:
- on def in first loop
// Assign register banks to ALL def registers on G_ instructions.
- on use in second loop
// Reassign uses of G_ instructions to only have register banks.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeRules.cpp:
```diff
@@ -0,0 +1,334 @@
+  for (unsigned i = 0; i < OpUniformityAndTypes.size(); ++i) {
+    if (OpUniformityAndTypes[i] == _) {
+      if (MI.getOperand(i).isReg() &&
+          MI.getOperand(i).getReg() != AMDGPU::NoRegister)
```
Avoid using NoRegister

petar-avramovic replied:
Removed NoRegister check since we don't need it at the moment.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeRules.cpp:
```diff
@@ -0,0 +1,334 @@
+    if (Slot != -1) {
+      if (MUI.isUniform(Reg))
+        return Uni[Slot];
+      else
```
No else after return

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeRules.cpp:
```diff
@@ -0,0 +1,334 @@
+  // Note: if fast rules are enabled, RegBankLLTMapping must be added in each
+  // slot that could "match fast Predicate". If not, Invalid Mapping is
+  // returned which results in failure, does not search "Slow Rules".
+  if (FastTypes != No) {
```
"No"?

petar-avramovic replied:
Renamed to NoFastRules

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeRules.cpp:
```diff
@@ -0,0 +1,334 @@
+int SetOfRulesForOpcode::getFastPredicateSlot(
+    UniformityLLTOpPredicateID Ty) const {
+  switch (FastTypes) {
+  case Standard:
```
Add braces

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeRules.h:
```diff
@@ -0,0 +1,258 @@
+} // end namespace AMDGPU
+} // end namespace llvm
+
+#endif
```
Missing end of file newline

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeRules.h:
```diff
@@ -0,0 +1,258 @@
+//===- AMDGPURBLegalizeRules -------------------------------------*- C++ -*-==//
```
Is it possible to share with the existing legalize rules?

petar-avramovic replied:
Don't think sharing is good option for this patch. RBLegalizeRules are much more flexible and I would assume faster because of "FastPredicateSlot".
If we add more IDs that would work with LLTs only we could rewrite Legalizer using RBLegalizeRules. Other way around is questionable, did not consider upgrading LegalityPredicate and LegalizeMutation to work with Register banks

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalize.cpp:
```diff
@@ -69,6 +81,241 @@ FunctionPass *llvm::createAMDGPURBLegalizePass() {
+      if (MI.getOpcode() == COPY) {
+        Register Dst = MI.getOperand(0).getReg();
+        Register Src = MI.getOperand(1).getReg();
+        if (!Dst.isVirtual() || !Src.isVirtual())
+          continue;
```
RBLegalize skips those copies, since it works with register banks only. Physical registers have to be copied to/from virtual register before being used by G_ instructions.
Also getVRegDef does not work for phys reg.

arsenm replied:
No G_* instruction can use a physical register

petar-avramovic replied:
Dst and Src are operands of a COPY, wanted to skip physreg copies

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -107,3 +107,183 @@ void IntrinsicLaneMaskAnalyzer::findLCSSAPhi(Register Reg) {
+  Register Dst = RFL->getOperand(0).getReg();
+  Register Src = RFL->getOperand(1).getReg();
+  MachineRegisterInfo &MRI = *B.getMRI();
+  if (!MRI.getRegBankOrNull(Dst))
+    MRI.setRegBank(Dst, RBI.getRegBank(SGPRRegBankID));
+  if (!MRI.getRegBankOrNull(Src))
+    MRI.setRegBank(Src, RBI.getRegBank(VGPRRegBankID));
```
Probably should add a contrainRegBank method to MRI, similar to constrainRegClass for this pattern

petar-avramovic replied:
How should it work in regards to possibility to insert illegal sgpr to vgpr copy and can it fail like register class version?
Or are we looking for something much simpler:
no reg bank -set reg bank
same reg bank - do nothing
different reg bank - insert copy

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalize.cpp:
```diff
@@ -69,6 +81,241 @@ FunctionPass *llvm::createAMDGPURBLegalizePass() {
+
+        MachineInstr *Trunc = MRI.getVRegDef(Src);
+        if (Trunc->getOpcode() != G_TRUNC)
+          continue;
```
Need to also capture Trunc MachineInstr to delete it if matched.

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalize.cpp:
```diff
@@ -69,6 +81,241 @@ FunctionPass *llvm::createAMDGPURBLegalizePass() {
+  static std::mutex GlobalMutex;
+  static SmallDenseMap<unsigned, std::unique_ptr<RegBankLegalizeRules>>
+      CacheForRuleSet;
+  std::lock_guard<std::mutex> Lock(GlobalMutex);
```
Never changes use lock_guard+mutex, why switch to call_once?

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalize.cpp:
```diff
@@ -69,6 +81,241 @@ FunctionPass *llvm::createAMDGPURBLegalizePass() {
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+
+  // Setup the instruction builder with CSE.
+  std::unique_ptr<MachineIRBuilder> MIRBuilder;
```
I guess that we can avoid it if we assume that TPC.isGISelCSEEnabled() is true. Copied from other passes.

petar-avramovic commented:
Rebase for new-reg-bank-select taking over AMDGPURegBankSelect
TODO: helper changes after DstOp improvements as it accepts bank+LLT

petar-avramovic commented:
Ping. There were changes because of improvements to builder accepting regbank+LLT, think I addressed most of the comments.

arsenm commented on llvm/lib/CodeGen/GlobalISel/MachineIRBuilder.cpp:
```diff
@@ -697,6 +697,15 @@ MachineInstrBuilder MachineIRBuilder::buildUnmerge(LLT Res,
+MachineIRBuilder::buildUnmerge(MachineRegisterInfo::VRegAttrs Attrs,
+                               const SrcOp &Op) {
+  LLT OpTy = Op.getLLTTy(*getMRI());
+  unsigned NumRegs = OpTy.getSizeInBits() / Attrs.Ty.getSizeInBits();
```
divideCoefficientBy? Does unmerge on scalable vectors work?

petar-avramovic replied:
divideCoefficientBy works only for vectors, we need to unmerge scalars also. Don't know about scalable vectors, they seem to only be used to say something is legal, don't know if there is actual lowering done for them. Other places in builder don't check for them.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -217,6 +217,74 @@ bool AMDGPUInstructionSelector::selectCOPY(MachineInstr &I) const {
   return true;
 }
 
+bool AMDGPUInstructionSelector::selectCOPY_SCC_VCC(MachineInstr &I) const {
```
I'm wondering if we should just avoid this situation

petar-avramovic replied:
This allows for more registers to be allocated to sgpr.
Avoiding them could be done later in some optimization pass when we can decide if we want to do global-isel equivalent of moveToVALU, but we don't do it for correctness but to influence register allocation to sgpr or vgpr.
Why would we want to avoid vcc to scc copy? Question for @nhaehnle also.

nhaehnle replied:
I don't think it's trivial to avoid this a priori. I agree that a separate cleanup optimization could do it, in any case I'd say it's best left to a separate focused change.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -217,6 +217,74 @@ bool AMDGPUInstructionSelector::selectCOPY(MachineInstr &I) const {
+  BuildMI(*BB, &I, DL, TII.get(AMDGPU::COPY), DstReg).addReg(AMDGPU::SCC);
+
+  I.eraseFromParent();
+  return RBI.constrainGenericRegister(DstReg, AMDGPU::SGPR_32RegClass, *MRI);
```
SReg_32?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp:
```diff
@@ -69,11 +82,297 @@ FunctionPass *llvm::createAMDGPURegBankLegalizePass() {
+
+static bool hasSgprS1(MachineFunction &MF, MachineRegisterInfo &MRI,
+                      const RegisterBank *SgprRB) {
+  LLT S1 = LLT::scalar(1);
```
const

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp:
```diff
@@ -69,11 +82,297 @@ FunctionPass *llvm::createAMDGPURegBankLegalizePass() {
+        if (MRI.getRegBankOrNull(Reg) != SgprRB || MRI.getType(Reg) != S1)
+          continue;
+
+        MI.getParent()->dump();
+        MI.dump();
```
Leftover debug

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp:
```diff
@@ -69,11 +82,297 @@ FunctionPass *llvm::createAMDGPURegBankLegalizePass() {
+  }
+};
+
+static bool hasSgprS1(MachineFunction &MF, MachineRegisterInfo &MRI,
```
Should move this to the machine verifier? 

Also this needs documentation and will get the unused warning in release build

petar-avramovic replied:
This is target specific, also should only be run between RegBankLegalize and instruction select not sure how to do that in machine verifier. Added [[maybe_unused]] for silence the compilers.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp:
```diff
@@ -69,11 +82,297 @@ FunctionPass *llvm::createAMDGPURegBankLegalizePass() {
+                      const RegisterBank *SgprRB) {
+  LLT S1 = LLT::scalar(1);
+  for (MachineBasicBlock &MBB : MF) {
+    for (MachineInstr &MI : make_early_inc_range(MBB)) {
```
Could loop over all virtual registers instead

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -0,0 +1,409 @@
+  switch (ID) {
+  case Vcc:
+    return VccRB;
+
```
Remove blank lines?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -0,0 +1,409 @@
+      MI.dump();
+      llvm_unreachable("missing fast rule for MI");
+    }
+
```
Remove blank

petar-avramovic commented:
Added LLVM_DEBUG around dumps, RB->RegBank rename for methods and class names

arsenm commented on llvm/include/llvm/CodeGen/MachineRegisterInfo.h:
```diff
@@ -674,6 +674,14 @@ class MachineRegisterInfo {
+  /// Return the register bank of \p Reg.
+  /// This shouldn't be used directly unless \p Reg has a register bank.
+  const RegisterBank *getRegBank(Register Reg) const {
+    assert(isa<const RegisterBank *>(VRegInfo[Reg.id()].first) &&
+           "Register bank not set, wrong accessor");
```
Assert redundant with the `cast<>`

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -106,3 +106,59 @@ void IntrinsicLaneMaskAnalyzer::findLCSSAPhi(Register Reg) {
+static LLT getReadAnyLaneSplitTy(LLT Ty) {
+  if (Ty.isVector()) {
+    LLT ElTy = Ty.getElementType();
+    if (ElTy == LLT::scalar(16))
+      return LLT::fixed_vector(2, 16);
```
```suggestion
    if (ElTy.getSizeInBits() == 16)
      return LLT::fixed_vector(2, EltTy);
```
Future proofing for float LLT

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -106,3 +106,59 @@ void IntrinsicLaneMaskAnalyzer::findLCSSAPhi(Register Reg) {
+  return LLT::scalar(32);
+}
+
+static Register buildReadAnyLane(MachineIRBuilder &B, Register VgprSrc,
```
Move the function body to avoid forward declaring

petar-avramovic replied:
there is circular dependency between buildReadAnyLane and unmergeReadAnyLane

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -106,3 +106,59 @@ void IntrinsicLaneMaskAnalyzer::findLCSSAPhi(Register Reg) {
+                              Register VgprSrc, const RegisterBankInfo &RBI) {
+  LLT Ty = B.getMRI()->getType(VgprSrc);
+  if (Ty.getSizeInBits() == 32) {
+    B.buildInstr(AMDGPU::G_READANYLANE, {SgprDst}, {VgprSrc});
```
This should have a G_AMDGPU prefix

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -249,7 +317,21 @@ bool AMDGPUInstructionSelector::selectPHI(MachineInstr &I) const {
-  // TODO: Verify that all registers have the same bank
+  // If inputs have register bank, assign corresponding reg class.
+  // Note: registers don't need to have the same reg bank.
+  for (unsigned i = 1; i < I.getNumOperands(); i += 2) {
```
```suggestion
  for (unsigned i = 1; I != I.getNumOperands(); i += 2) {
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp:
```diff
@@ -69,11 +82,295 @@ FunctionPass *llvm::createAMDGPURegBankLegalizePass() {
+    return RC && TRI.isSGPRClass(RC) && MRI.getType(Reg) == LLT::scalar(1);
+  }
+
+  void cleanUpAfterCombine(MachineInstr &MI, MachineInstr *Optional0) {
```
static

petar-avramovic replied:
this is member function of AMDGPURegBankLegalizeCombner since it needs access to MRI

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp:
```diff
@@ -69,11 +82,295 @@ FunctionPass *llvm::createAMDGPURegBankLegalizePass() {
+      Optional0->eraseFromParent();
+  }
+
+  std::pair<MachineInstr *, Register> tryMatch(Register Src, unsigned Opcode) {
```
Why this instead of using mi_match?

petar-avramovic replied:
This is much more convenient, shorter and easier to read.
for mi_match need to declare MI and Reg to pass as arguments, then need to match same thing twice using something like m_all_of(m_MInstr(...), m_Reg(...))).
This is a simple opcode check. Also when written like this it allows for structured binding of MI and Reg

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp:
```diff
@@ -69,11 +82,295 @@ FunctionPass *llvm::createAMDGPURegBankLegalizePass() {
+  const LLT S1 = LLT::scalar(1);
+  for (unsigned i = 0; i < MRI.getNumVirtRegs(); ++i) {
+    Register Reg = Register::index2VirtReg(i);
+    if (MRI.def_empty(Reg))
+      continue;
+
+    if (MRI.getType(Reg) != S1)
+      continue;
```
```suggestion
    if (MRI.getType(Reg) != S1 || MRI.def_empty(Reg))
      continue;
```

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.h:
```diff
@@ -9,7 +9,11 @@
 #ifndef LLVM_LIB_TARGET_AMDGPU_AMDGPUGLOBALISELUTILS_H
 #define LLVM_LIB_TARGET_AMDGPU_AMDGPUGLOBALISELUTILS_H
 
+#include "AMDGPURegisterBankInfo.h"
+#include "MCTargetDesc/AMDGPUMCTargetDesc.h"
 #include "llvm/ADT/DenseSet.h"
+#include "llvm/CodeGen/GlobalISel/MIPatternMatch.h"
+#include "llvm/CodeGen/GlobalISel/MachineIRBuilder.h"
```
We should avoid unnecessary `#include`s in headers (for compile time reason). I doubt that all of these are really needed.

petar-avramovic replied:
Updated. Are we ready to merge the whole stack? This commit is still missing approval.

nhaehnle reviewed (commented):
LGTM aside from a small nit

petar-avramovic committed:
AMDGPU/GlobalISel: AMDGPURegBankLegalize

Lower G_ instructions that can't be inst-selected with register bank
assignment from AMDGPURegBankSelect based on uniformity analysis.
- Lower instruction to perform it on assigned register bank
- Put uniform value in vgpr because SALU instruction is not available
- Execute divergent instruction in SALU - "waterfall loop"

Given LLTs on all operands after legalizer, some register bank
assignments require lowering while other do not.
Note: cases where all register bank assignments would require lowering
are lowered in legalizer.

AMDGPURegBankLegalize goals:
- Define Rules: when and how to perform lowering
- Goal of defining Rules it to provide high level table-like brief
  overview of how to lower generic instructions based on available
  target features and uniformity info (uniform vs divergent).
- Fast search of Rules, depends on how complicated Rule.Predicate is
- For some opcodes there would be too many Rules that are essentially
  all the same just for different combinations of types and banks.
  Write custom function that handles all cases.
- Rules are made from enum IDs that correspond to each operand.
  Names of IDs are meant to give brief description what lowering does
  for each operand or the whole instruction.
- AMDGPURegBankLegalizeHelper implements lowering algorithms

Since this is the first patch that actually enables -new-reg-bank-select
here is the summary of regression tests that were added earlier:
- if instruction is uniform always select SALU instruction if available
- eliminate back to back vgpr to sgpr to vgpr copies of uniform values
- fast rules: small differences for standard and vector instruction
- enabling Rule based on target feature - salu_float
- how to specify lowering algorithm - vgpr S64 AND to S32
- on G_TRUNC in reg, it is up to user to deal with truncated bits
  G_TRUNC in reg is treated as no-op.
- dealing with truncated high bits - ABS S16 to S32
- sgpr S1 phi lowering
- new opcodes for vcc-to-scc and scc-to-vcc copies
- lowering for vgprS1-to-vcc copy (formally this is vgpr-to-vcc G_TRUNC)
- S1 zext and sext lowering to select
- uniform and divergent S1 AND(OR and XOR) lowering - inst-selected into
  SALU instruction
- divergent phi with uniform inputs
- divergent instruction with temporal divergent use, source instruction
  is defined as uniform(AMDGPURegBankSelect) - missing temporal
  divergence lowering
- uniform phi, because of undef incoming, is assigned to vgpr. Will be
  fixed in AMDGPURegBankSelect via another fix in machine uniformity
  analysis.

petar-avramovic commented:
### Merge activity

* **Jan 24, 6:10 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/112864).
